package/MDAnalysis/analysis/base.py

```diff
@@ -238,17 +239,29 @@ def _setup_frames(self, trajectory, start=None, stop=None, step=None):
             stop frame of analysis
         step : int, optional
             number of frames to skip between each analysed frame
+        frames : array_like, optional
+            array of integers or booleans to slice trajectory
 
 
         .. versionchanged:: 1.0.0
             Added .frames and .times arrays as attributes
+
+        .. versionchanged:: 2.1.0
+            Added frames processing
         """
         self._trajectory = trajectory
-        start, stop, step = trajectory.check_slice_indices(start, stop, step)
+        if frames is not None:
+            if any([start, stop, step]):
+                raise ValueError("start/stop/step cannot be combined with frames")
+            slicer = frames
+        else:
+            start, stop, step = trajectory.check_slice_indices(start, stop, step)
+            slicer = slice(start, stop, step)
+        self._sliced_trajectory = trajectory[slicer]
         self.start = start
         self.stop = stop
         self.step = step
-        self.n_frames = len(range(start, stop, step))
+        self.n_frames = len(self._sliced_trajectory)
         self.frames = np.zeros(self.n_frames, dtype=int)
```

Comment:
I can see this becoming confusing real quickly, we're offering `frames` as an input argument to run and then offering a different thing contained in the `frames` user-facing variable. Thoughts?

Maybe we should call the input argument `indices` instead?

Reply:
I see your point. Although in principle `self.frames` should be equivalent to what the user inputs in `frames` in the sense that running a second time `a.run(frames=a.frames)` should give the identical answer (assuming that `ts.frame` is the true frame index), which I like. If the input `frames` was a boolean array or a `slice()` then `self.frames` will be an int array of frame indices but _should_ be functionally equivalent.

I am not too fond of calling the input `frames` kwarg `indices` because (I think) we typically use it for indexing atoms. (?)

Reply:
It's a little bit messy but I'm not too fussed.

I'm not sure what (if anything) would need adding, but do we want to alter the attributes description for frame to make this clearer?

Reply:
@IAlibay How should I alter the attribute description for `frames`? 
It is currently "array of Timestep frame indices. Only exists after calling AnalysisBase.run"
Something like "array of Timestep frame indices. Only exists after calling AnalysisBase.run. If the 'frames' 
option is used in .run(), the input for 'frames' will be returned"

Reply:
With hindsight, let's leave the attribute as it is. I'm not really sure how to amend (or if it's at all possible) it without making it more complex.